File: backend/app/schemas/vehicle.py
from datetime import datetime

from pydantic import (
    BaseModel,
    ConfigDict,
    Field,
    field_validator,
)

from app.schemas.mot import MotTestRead


def normalise_registration(value: str) -> str:
    if not isinstance(value, str):
        raise ValueError("Registration must be a string")

    registration = "".join(
        value.upper().split()
    )

    if not registration:
        raise ValueError("Registration cannot be empty")

    if len(registration) > 8:
        raise ValueError("Registration is too long")

    if not registration.isalnum():
        raise ValueError(
            "Registration must only contain letters and numbers"
        )

    return registration


class VehicleRegistration(BaseModel):
    registration: str = Field(
        min_length=1,
        max_length=8,
    )

    @field_validator(
        "registration",
        mode="before",
    )
    @classmethod
    def validate_registration(
        cls,
        value: str,
    ) -> str:
        return normalise_registration(value)


class VehicleBase(VehicleRegistration):
    make: str = Field(
        min_length=1,
        max_length=100,
    )

    model: str = Field(
        min_length=1,
        max_length=100,
    )

    fuel_type: str | None = Field(
        default=None,
        max_length=50,
    )

    engine_size: int | None = Field(
        default=None,
        gt=0,
    )

    colour: str | None = Field(
        default=None,
        max_length=50,
    )

    year: int | None = Field(
        default=None,
        ge=1886,
        le=2100,
    )


class VehicleCreate(VehicleBase):
    pass


class VehicleLookupRequest(VehicleRegistration):
    pass


class VehicleRead(VehicleBase):
    id: int
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(
        from_attributes=True,
    )


class VehicleImportResponse(BaseModel):
    vehicle: VehicleRead
    mot_tests_found: int
    mot_tests_saved: int


class VehicleCheckResponse(VehicleBase):
    mot_tests_found: int
    mot_tests: list[MotTestRead]

    in_garage: bool
    garage_vehicle_id: int | None = None

File: frontend/src/types/vehicle.ts
export interface MotDefect {
  text: string;
  type: string;
  dangerous: boolean;
}


export interface MotTest {
  mot_test_number: string;
  completed_at: string;

  data_source: string | null;
  expiry_date: string | null;
  registration_at_time_of_test: string | null;

  test_result: string | null;

  odometer_value: number | null;
  odometer_unit: string | null;
  odometer_result_type: string | null;

  defects: MotDefect[];
}


export interface VehicleCheckResponse {
  registration: string;
  make: string;
  model: string;

  fuel_type: string | null;
  engine_size: number | null;
  colour: string | null;
  year: number | null;

  mot_tests_found: number;
  mot_tests: MotTest[];

  in_garage: boolean;
  garage_vehicle_id: number | null;
}


export interface GarageVehicle {
  id: number;

  registration: string;
  make: string;
  model: string;

  fuel_type: string | null;
  engine_size: number | null;
  colour: string | null;
  year: number | null;

  created_at: string;
  updated_at: string;
}


export interface VehicleImportResponse {
  vehicle: GarageVehicle;

  mot_tests_found: number;
  mot_tests_saved: number;
}


export interface MotHistoryRefreshResponse {
  vehicle_id: number;
  registration: string;

  mot_tests_found: number;
  mot_tests_saved: number;
}